File: pqkmeans/encoder/encoder_sample.py
import _pqkmeans
import typing
from .encoder_base import EncoderBase


class EncoderSample(EncoderBase):
    def __init__(self):
        self._impl = _pqkmeans.EncoderSample()

    def fit_generator(self, x_train):
        # type: (typing.Iterable[typing.Iterable[float]]) -> None
        self._impl.fit_generator(x_train)

    def transform_generator(self, x_test):
        # type: (typing.Iterable[typing.Iterator[float]]) -> Any
        for vector in x_test:
            yield self._impl.transform_one(vector)

    def inverse_transform_generator(self, x_test):
        # type: (typing.Iterable[typing.Iterator[int]]) -> Any
        for vector in x_test:
            yield self._impl.inverse_transform_one(vector)

File: src/encoder/encoder_sample.cpp
#include "encoder_sample.h"

namespace pqkmeans {
void EncoderSample::fit_generator(py::iterator iterator) {
    auto begin = std::begin(iterator);
    auto end = std::end(iterator);
    for (auto itr = begin; itr != end; ++itr) {
        auto item = (*itr).cast<std::vector<float>>();
        auto found = std::find(index_dict.begin(), index_dict.end(), item);

        if (found == index_dict.end()) {
            index_dict.push_back(item);
        }
    }
}

std::vector<long> EncoderSample::transform_one(const std::vector<float> &vector) {
    auto found = std::find(index_dict.begin(), index_dict.end(), vector);
    if (found != index_dict.end()) {
        std::vector<long> ret_vector;
        ret_vector.push_back(std::distance(index_dict.begin(), found));
        return ret_vector;
    } else {
        throw std::invalid_argument("couldn't handle input vector");
    }
}

std::vector<float> EncoderSample::inverse_transform_one(const std::vector<long> &vector) {
    std::size_t value = (std::size_t) vector[0];
    if (value < index_dict.size()) {
        return index_dict[value];
    } else {
        throw std::invalid_argument("couldn't handle input vector");
    }
}
}
